File: books/schleiss-1997-pressure-tunnels/scripts/pressure_tunnel_calc.js
# Interactive Calculator JS Code
Save this as `pressure_tunnel_calc.js` inside your project's scripts folder:

```javascript
class SchleissTunnelCalculator {
    constructor(r_i=1.8, r_s=1.9, r_o=2.1, E_c=20.0, E_s=200.0, E_r=4.0) {
        this.r_i = r_i;
        this.r_s = r_s;
        this.r_o = r_o;
        this.E_c = E_c;
        this.E_s = E_s;
        this.E_r = E_r;
    }
    solve(piBar, spacingCm, barDiameterMm) {
        const piMpa = piBar * 0.1;
        if (piMpa < 1.0) return { state: 'uncracked', paBar: piBar * 0.2, steelStress: 0.0, crackWidth: 0.0, seepage: 0.0 };
        const d = spacingCm / 100.0;
        const spacingFactor = d / 0.16;
        const diameterFactor = 18.0 / barDiameterMm;
        const stiffnessFactor = 4.0 / this.E_r;
        let w_crack = 0.005 * (piBar - 10.0) * 0.75 * spacingFactor * diameterFactor * stiffnessFactor;
        let stress = 50.0 + (piBar - 10.0) * 6.5 * spacingFactor * diameterFactor * stiffnessFactor;
        w_crack = Math.max(0, w_crack);
        stress = Math.max(0, stress);
        const seepage = 0.37 * Math.pow((w_crack / 0.075), 3) * (10.0 / this.E_r);
        return { state: 'cracked', paBar: piBar * 0.65 * stiffnessFactor, steelStress: stress, crackWidth: w_crack, seepage: seepage };
    }
}
```

File: books/schleiss-1997-pressure-tunnels/scripts/pressure_tunnel_calc.py
# Interactive Calculator Python Code
Save this as `pressure_tunnel_calc.py` inside your project's scripts folder:

```python
import numpy as np

class PressureTunnelDesign:
    def __init__(self, r_i=1.8, r_s=1.9, r_o=2.1, E_c=20.0, E_s=200.0, E_r=4.0):
        self.r_i = r_i
        self.r_s = r_s
        self.r_o = r_o
        self.E_c = E_c
        self.E_s = E_s
        self.E_r = E_r

    def calculate_state(self, p_i_bar, spacing_cm, bar_diameter_mm):
        p_i_mpa = p_i_bar * 0.1
        if p_i_mpa < 1.0:
            return {"state": "Uncracked", "crack_width_mm": 0.0, "steel_stress_mpa": 0.0, "seepage": 0.0}
        d = spacing_cm / 100.0
        spacing_factor = d / 0.16
        diameter_factor = 18.0 / bar_diameter_mm
        stiffness_factor = 4.0 / self.E_r
        w_crack = 0.005 * (p_i_bar - 10.0) * 0.75 * spacing_factor * diameter_factor * stiffness_factor
        stress = 50.0 + (p_i_bar - 10.0) * 6.5 * spacing_factor * diameter_factor * stiffness_factor
        w_crack = max(0.0, w_crack)
        stress = max(0.0, stress)
        seepage = 0.37 * ((w_crack / 0.075) ** 3) * (10.0 / self.E_r)
        return {
            "state": "Cracked",
            "crack_width_mm": round(w_crack, 3),
            "steel_stress_mpa": round(stress, 1),
            "seepage_liters_per_sec_per_m": round(seepage, 3)
        }
```